File: frontend/src/types/deck.ts
// Deck system types

export type AccentType = 'heiban' | 'atamadaka' | 'nakadaka' | 'odaka';

export interface Card {
  id: string;
  word: string;
  reading: string;
  meaning?: string;
  accent_position: number;
  accent_type: AccentType;
  pitch_pattern: string;
  audio_url?: string;
  pair_word?: string;
  notes?: string;
  sort_order: number;
}

export interface DeckSummary {
  id: string;
  slug: string;
  title: string;
  title_ja?: string;
  description?: string;
  phase: number;
  is_free: boolean;
  card_count: number;
  // User progress
  cards_seen: number;
  cards_mastered: number;
  is_started: boolean;
  is_completed: boolean;
  progress_percent: number;
}

export interface DeckDetail {
  id: string;
  slug: string;
  title: string;
  title_ja?: string;
  description?: string;
  phase: number;
  is_free: boolean;
  card_count: number;
  cards: Card[];
  last_card_index: number;
  cards_seen: number;
}

export interface DeckListResponse {
  decks: DeckSummary[];
  total_cards: number;
  cards_learned: number;
}

export interface LearningStats {
  total_cards_seen: number;
  total_cards_mastered: number;
  decks_started: number;
  decks_completed: number;
  cards_due_today: number;
}

// Accent type display names
export const ACCENT_TYPE_LABELS: Record<AccentType, string> = {
  heiban: '平板',
  atamadaka: '頭高',
  nakadaka: '中高',
  odaka: '尾高',
};

// Phase labels
export const PHASE_LABELS: Record<number, string> = {
  1: 'Foundation',
  2: 'Core Vocabulary',
  3: 'Structure',
  4: 'Fluency',
};

File: backend/app/models/deck.py
"""Deck and Card models for the learning system."""

from datetime import datetime
from pydantic import BaseModel


class Card(BaseModel):
    """A single learning card."""

    id: str
    word: str
    reading: str
    meaning: str | None = None
    accent_position: int
    accent_type: str  # heiban, atamadaka, nakadaka, odaka
    pitch_pattern: str  # LH, HL, LHL, etc.
    audio_url: str | None = None
    pair_word: str | None = None
    notes: str | None = None
    sort_order: int = 0


class DeckSummary(BaseModel):
    """Deck summary for library listing."""

    id: str
    slug: str
    title: str
    title_ja: str | None = None
    description: str | None = None
    phase: int
    is_free: bool
    card_count: int
    # User progress (populated if authenticated)
    cards_seen: int = 0
    cards_mastered: int = 0
    is_started: bool = False
    is_completed: bool = False
    progress_percent: int = 0


class DeckDetail(BaseModel):
    """Full deck with cards."""

    id: str
    slug: str
    title: str
    title_ja: str | None = None
    description: str | None = None
    phase: int
    is_free: bool
    card_count: int
    cards: list[Card]
    # User progress
    last_card_index: int = 0
    cards_seen: int = 0


class DeckListResponse(BaseModel):
    """Response for deck listing."""

    decks: list[DeckSummary]
    total_cards: int = 0
    cards_learned: int = 0


class ProgressUpdate(BaseModel):
    """Request to update deck progress."""

    card_index: int
    card_id: str | None = None
    seen: bool = True
    mastered: bool = False


class CardReviewResult(BaseModel):
    """Result of reviewing a card (for SRS)."""

    card_id: str
    result: str  # correct, incorrect, hard, easy


class ReviewDueResponse(BaseModel):
    """Cards due for review."""

    cards: list[Card]
    total_due: int
    estimated_minutes: int
